File: notebook/july_evaluation/Scenario3/paper_code/SEIRV_fit_tempfcn.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% This code carries out data fitting for the second wave using the SEIRV
% model. 
%
% NOTE: High titer -> Mean half-life 0.99 days
%       Low titer  -> Mean half-life 7.9 days
%       These values impact tau0 in the getDecay() function.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

rng('default');

clear;clc
set(groot, 'defaultAxesTickLabelInterpreter','latex');
set(groot, 'defaultLegendInterpreter','latex');
set(0,'defaulttextInterpreter','latex','defaultAxesFontSize',16) 
format long

bl = '#0072BD';
br = '#D95319';

%% Load data
load('data.mat')

V = cRNA2.*F2;
split = 78; %split 1 week after first day of vaccine (12/11/2020)
V = V(1:split);
tspan = 1:length(V);

%% Curve-fitting
options = optimoptions('fmincon','TolX',1e-12,'TolFun',1e-12,'MaxIter',50000,'MaxFunEvals',100000,'display','off');

  beta_fixed = 4.48526e7;
    lb = [0     51    beta_fixed  10 ];
    ub = [1E-4  796   beta_fixed  5000];
    p0 = [9.06e-08 360 beta_fixed 1182];

% 
%% try Global Search
gs = GlobalSearch;
ms = MultiStart('Display','iter');

problem = createOptimProblem('fmincon','x0',p0,...
    'objective',@(param)obj_fun(param,tspan,V),'lb',lb,'ub',ub);

[best_params,SSE] = run(ms,problem,25);

parameter = ["lambda";"alpha";"beta";"E(0)";"SSE"];
estimated_val = [best_params';SSE];
t = table(parameter,estimated_val)

%% Simulate with best params

alpha = best_params(2);
beta = best_params(3);

traveltime = 18; % hours
k = getDecay(1); % use first time point

eta = 1 - exp(-k*traveltime);

% total population served by DITP
N0 = 2300000;

E0 = best_params(4);
I0 = V(1)/(alpha*beta*(1-eta));
R0 = 0;
S0 = N0 - (E0 + I0 + R0);
V0 = V(1); % use first data point 
ICs  = [S0 E0 I0 R0 V0 E0];

[T,Y] = ode45(@SEIRV,1:length(cRNA2),ICs,[],best_params);


%% Plot
time = datetime(2020,9,30) + caldays(0:length(cRNA2)-1);

figure()
    t = tiledlayout(1,2);

    nexttile
    plot(time2(2:end),log10(diff(Y(:,5))),'LineWidth',2); hold on
    plot(time2(2:end),log10(cRNA2(2:end).*F2(2:end)),'.','markersize',20,'LineWidth',2,'Color',br);
    ylabel('$\log_{10}$ viral RNA copies')
    xline(split,'--','LineWidth',2,'Color',[0 1 0])
    ylim([13.5 inf])
    xlim([time(18-1) time(116-1)])


    nexttile
    plot(time(2:end),log10(diff(Y(:,6))),'LineWidth',2); hold on
    p2 = plot(time2(2:end),log10(newRepCases2(2:end)),'LineWidth',2,'Color',br);
    ylabel('$\log_{10}$ Daily Incidence');

    [max1,index1] = max(diff(Y(:,6))); %simulation max
    xline(time2(index1+1),'--','LineWidth',2,'Color',bl)
    [max2,index2] = max(newRepCases2); %simulation max
    xline(time2(index2),'--','LineWidth',2,'Color',br)

    legend('Model','Data','Location','NorthWest')
    ylim([2.379 4.5])
    xlim([time2(2) time2(118)])

    hold off
    
    %%
    
    f = gcf;
    exportgraphics(f,'fitting_with_temperature.pdf','Resolution',600)

    %%
    
    figure
    box on; hold on;

    %estimate R
    y = (diff(Y(:,6)));
    x = (newRepCases2(2:end));
    X = [ones(length(x),1) x];
    b = X\y;

    yCalc2 = X*b;%b1*x;
    scatter(x,y,20,'k','LineWidth',2);
    plot(x,yCalc2,'r','LineWidth',2)
    ylim([0 inf])

    ylabel('Predicted cases');
    xlabel('Reported cases')

    %calculate R2
    Rsq2 = 1 - sum((y-yCalc2).^2)/sum((y-mean(y)).^2);

    R = corrcoef(x,y); 

    f = gcf;
    exportgraphics(f,'corr_1.pdf','Resolution',600)
%% functions

function err = obj_fun(param,tspan,data)
    traveltime = 18;% hours
    k = getDecay(1); % use first time point

    eta = 1 - exp(-k*traveltime);

    % total population served by DITP
    N0 = 2300000;

    E0 = param(4);
    I0 = data(1)/(param(2)*param(3)*(1-eta));
    R0 = 0;
    S0 = N0 - (E0 + I0 + R0);
    V0 = data(1);                
    ICs  = [S0 E0 I0 R0 V0 E0];

    [~,Y] = ode45(@SEIRV,tspan,ICs,[],param(1:4));

    % get daily virus
    cumVirus = Y(:,5);
    dailyVirus = diff(cumVirus);

    temp = log10(data(2:end)) - log10(abs(dailyVirus));
    adiff = rmmissing(temp);

    err = sum((adiff).^2);
end

function k = getDecay(t)
    % compute temperature-adjusted decay rate of viral RNA
    
    % high titer -> tau0 = 0.99 days * 24 hours/day = 23.76
    % low titer  -> tau0 = 7.9 days * 24 hours/day  = 189.6

    tau0 = 189.6;%23.76;
    Q0 = 2.5;
    T0 = 20;

    % get current temperature using best-fit sine function
    A = 3.624836409841919;
    B = 0.020222716119084;
    C = 4.466530666828714;
    D = 16.229757918464635;

    T = A*sin(B*t - C) + D;

    tau = tau0*Q0.^(-(T - T0)/10);

    k = log(2)./tau;

end




function dy = SEIRV(t,y,param)
    % parameters to be fit
    lambda = param(1);
    alpha = param(2);
    beta = param(3);

    dy = zeros(6,1);
    
    S = y(1);  
    E = y(2);      
    I = y(3);
    R = y(4);
    V = y(5);

    traveltime = 18; % hours
    k = getDecay(t);

    eta = 1 - exp(-k*traveltime);

    sigma = 1/3;
    gamma = 1/8;
    

    dy(1) = -lambda*S*I;
    dy(2) = lambda*S*I - sigma*E;                               
    dy(3) = sigma*E - gamma*I;
    dy(4) = gamma*I;
    dy(5) = alpha*beta*(1-eta)*I;
    dy(6) = lambda*S*I;       % track cumulative cases
end

File: notebook/july_evaluation/Scenario3/paper_code/SEIRV_fit_tempfcn.py
import numpy as np
import scipy.optimize as opt
from scipy.io import loadmat
from scipy.integrate import odeint
import pandas as pd

# Seed the random number generator
np.random.seed(0)

# Load data
data = loadmat('data.mat')
cRNA2 = np.array(data['cRNA2'])
F2 = np.array(data['F2'])
V = cRNA2 * F2
split = 78
V = V[:split]
tspan = np.arange(1, len(V) + 1)

# Curve-fitting
beta_fixed = 4.48526e7
lb = [0, 51, beta_fixed, 10]
ub = [1e-4, 796, beta_fixed, 5000]
p0 = [9.06e-08, 360, beta_fixed, 1182]

# Define objective function
from scipy.integrate import solve_ivp
import numpy as np

def obj_fun(param, tspan, data):
    """"""
    traveltime = 18 # hours
    k = getDecay(1) # use first time point

    eta = 1 - np.exp(-k*traveltime)

    # total population served by DITP
    N0 = 2300000

    E0 = param[3]
    I0 = data[0] / (param[1] * param[2] * (1-eta))
    R0 = 0
    S0 = N0 - (E0 + I0 + R0)
    V0 = data[0] # use first data point
    ICs  = [S0, E0, I0, R0, V0, E0]

    # You need to have the function SEIRV defined somewhere, as you did in MATLAB
    sol = solve_ivp(lambda t, y: SEIRV(t, y, param[0:3]), [tspan[0], tspan[-1]], ICs)

    # get daily virus
    cumVirus = sol.y[4]
    dailyVirus = np.diff(cumVirus)

    temp = np.log10(data[1:]) - np.log10(np.abs(dailyVirus))
    adiff = temp[~np.isnan(temp)]

    err = np.sum((adiff)**2)
    return err


# Perform optimization
bounds = opt.Bounds(lb, ub)
result = opt.differential_evolution(lambda x: obj_fun(x, tspan, V), bounds=bounds)

best_params = result.x
SSE = result.fun

# Create data frame with parameters and their estimated values
parameter = ["lambda", "alpha", "beta", "E(0)", "SSE"]
estimated_val = np.append(best_params, SSE)
df = pd.DataFrame({'parameter': parameter, 'estimated_val': estimated_val})
print(df)

# Simulate with best params
alpha = best_params[1]
beta = best_params[2]

traveltime = 18  # hours
k = getDecay(1)  # use first time point
eta = 1 - np.exp(-k * traveltime)
N0 = 2300000  # total population served by DITP
E0 = best_params[3]
I0 = V[0] / (alpha * beta * (1 - eta))
R0 = 0
S0 = N0 - (E0 + I0 + R0)
V0 = V[0]
ICs = [S0, E0, I0, R0, V0, E0]

# Define SEIRV model function
def SEIRV(ICs, t, best_params):
    # Define SEIRV code here...

T, Y = odeint(SEIRV, ICs, np.arange(1, len(cRNA2) + 1), args=(best_params,))

# Plot results
import matplotlib.pyplot as plt
import datetime
from scipy.stats import linregress

# Conversion of MATLAB's datetime to Python's datetime
start_date = datetime.datetime(2020, 9, 30)
time = [start_date + datetime.timedelta(days=i) for i in range(len(cRNA2))]

# Create plot
fig, axes = plt.subplots(1, 2, figsize=(10, 5))

# First subplot
axes[0].plot(time[1:], np.log10(np.diff(Y[:,4])), linewidth=2)
axes[0].scatter(time[1:], np.log10(cRNA2[1:] * F2[1:]), color='r', s=20, linewidth=2)
axes[0].axvline(x=time[split - 1], linestyle='--', linewidth=2, color='g')
axes[0].set_ylim([13.5, None])
axes[0].set_xlim([time[18 - 1], time[116 - 1]])
axes[0].set_ylabel(r'$\log_{10}$ viral RNA copies')

# Second subplot
axes[1].plot(time[1:], np.log10(np.diff(Y[:,5])), linewidth=2)
axes[1].plot(time[1:], np.log10(newRepCases2[1:]), linewidth=2, color='r')
axes[1].set_ylabel(r'$\log_{10}$ Daily Incidence')

index1 = np.argmax(np.diff(Y[:,5]))
index2 = np.argmax(newRepCases2)
axes[1].axvline(x=time[index1 + 1], linestyle='--', linewidth=2, color='b')
axes[1].axvline(x=time[index2], linestyle='--', linewidth=2, color='r')
axes[1].legend(['Model', 'Data'], loc='upper left')
axes[1].set_ylim([2.379, 4.5])
axes[1].set_xlim([time[1], time[118]])

plt.tight_layout()

# Save figure
plt.savefig('fitting_with_temperature.pdf', dpi=600)
plt.clf()

# Second figure
plt.figure()
plt.box(True)
y = np.diff(Y[:,5])
x = newRepCases2[1:]
X = np.column_stack((np.ones(len(x)), x))
b = np.linalg.lstsq(X, y, rcond=None)[0]
yCalc2 = X @ b

plt.scatter(x, y, color='k', s=20, linewidth=2)
plt.plot(x, yCalc2, color='r', linewidth=2)
plt.ylim([0, None])
plt.ylabel('Predicted cases')
plt.xlabel('Reported cases')

# Calculate R2
Rsq2 = 1 - np.sum((y - yCalc2)**2) / np.sum((y - np.mean(y))**2)
R = np.corrcoef(x, y)

plt.savefig('corr_1.pdf', dpi=600)




import numpy as np

def getDecay(t):
    # compute temperature-adjusted decay rate of viral RNA
    
    # high titer -> tau0 = 0.99 days * 24 hours/day = 23.76
    # low titer  -> tau0 = 7.9 days * 24 hours/day  = 189.6

    tau0 = 189.6 #23.76
    Q0 = 2.5
    T0 = 20

    # get current temperature using best-fit sine function
    A = 3.624836409841919
    B = 0.020222716119084
    C = 4.466530666828714
    D = 16.229757918464635

    T = A * np.sin(B * t - C) + D

    tau = tau0 * Q0 ** (-(T - T0) / 10)

    k = np.log(2) / tau

    return k

import numpy as np
from scipy.integrate import odeint

def SEIRV(y, t, param):
    # parameters to be fit
    lambda_ = param[0]
    alpha = param[1]
    beta = param[2]

    S, E, I, R, V, cumulative_cases = y

    traveltime = 18  # hours
    k = getDecay(t)

    eta = 1 - np.exp(-k*traveltime)

    sigma = 1/3
    gamma = 1/8

    dS = -lambda_ * S * I
    dE = lambda_ * S * I - sigma * E
    dI = sigma * E - gamma * I
    dR = gamma * I
    dV = alpha * beta * (1 - eta) * I
    dcumulative_cases = lambda_ * S * I  # track cumulative cases

    dy = [dS, dE, dI, dR, dV, dcumulative_cases]

    return dy

# Note: to use odeint in scipy to integrate this system of differential equations, you would do something like the following:
# y0 = [S0, E0, I0, R0, V0, cumulative_cases0]
# t = np.linspace(start_time, end_time, num_timepoints)
# result = odeint(SEIRV, y0, t, args=(param,))

